File: api/app/sim/geometry.py
"""Plane geometry, kept to operations IEEE-754 specifies exactly.

Add, subtract, multiply, divide and square root only — no trigonometry and no
fractional powers. Those are implementation-defined, and a battle that replays
differently on another interpreter build is not deterministic. `math.sqrt` maps
to the hardware instruction and is correctly rounded, so it is safe.
"""

from __future__ import annotations

import math

from app.sim.types import Vec2


def distance(a: Vec2, b: Vec2) -> float:
    dx = b.x - a.x
    dy = b.y - a.y
    return math.sqrt(dx * dx + dy * dy)


def move_toward(origin: Vec2, target: Vec2, step: float) -> Vec2:
    """Steps `origin` toward `target` by at most `step`, stopping exactly on it."""
    gap = distance(origin, target)
    if gap == 0 or step >= gap:
        return Vec2(target.x, target.y)

    scale = step / gap
    return Vec2(
        origin.x + (target.x - origin.x) * scale,
        origin.y + (target.y - origin.y) * scale,
    )

File: api/src/sim/geometry.ts
/**
 * Plane geometry, kept to the operations IEEE-754 specifies exactly — add,
 * subtract, multiply, divide, and square root. No trigonometry and no `**` with
 * a fractional exponent: those are implementation-defined, and a battle that
 * replays differently on another Node build is not deterministic.
 */
import type { Vec2 } from './types.js';

export function distance(a: Vec2, b: Vec2): number {
  const dx = b.x - a.x;
  const dy = b.y - a.y;
  return Math.sqrt(dx * dx + dy * dy);
}

/** Steps `from` toward `to` by at most `step`, stopping exactly on `to`. */
export function moveToward(from: Vec2, to: Vec2, step: number): Vec2 {
  const gap = distance(from, to);
  if (gap === 0 || step >= gap) {
    return { x: to.x, y: to.y };
  }
  const scale = step / gap;
  return { x: from.x + (to.x - from.x) * scale, y: from.y + (to.y - from.y) * scale };
}
